plugins/octant/cmd/antrea-octant-plugin/traceflow.go

```diff
@@ -88,292 +95,236 @@ func getDstType(tf *crdv1alpha1.Traceflow) string {
+		var srcLen int
+		var isSrcPodType bool
+		dstLen := len(dst)
 
-		tf := &crdv1alpha1.Traceflow{
-			ObjectMeta: v1.ObjectMeta{
-				Name: tfName,
-			},
-			Spec: crdv1alpha1.TraceflowSpec{
-				Source: crdv1alpha1.Source{
-					Namespace: srcNamespace,
-					Pod:       srcPod,
-				},
-				Destination: destination,
-				Packet: crdv1alpha1.Packet{
-					IPHeader: crdv1alpha1.IPHeader{
-						Protocol: crdv1alpha1.SupportedProtocols[protocol[0]],
-					},
-				},
-			},
+		// Judge the source type and get source according to the type.
+		srcType, err := request.Payload.StringSlice(srcTypeCol)
+		if err != nil || len(srcType) == 0 {
+			alertPrinter(request, invalidInputMsg+"failed to get srcType as string slice:",
+				"Invalid source type choice, please check your input and submit again.", nil, err)
+			return nil
+		}
+		src, err := request.Payload.String(srcCol)
+		if err != nil {
+			alertPrinter(request, invalidInputMsg+"failed to get src as string:",
+				"Failed to get source as string: ", nil, err)
+			return nil
+		}
+		srcLen = len(src)
+		if srcLen == 0 && dstLen == 0 {
+			alertPrinter(request, invalidInputMsg+"one of source/destination must be set, and must be a Pod",
+				"one of source/destination must be set, and must be a Pod, please check your input and submit again.", nil, nil)
+			return nil
 		}
 
-		switch tf.Spec.Packet.IPHeader.Protocol {
-		case crdv1alpha1.TCPProtocol:
-			{
-				tf.Spec.Packet.TransportHeader.TCP = &crdv1alpha1.TCPHeader{
-					Flags: 2,
-				}
-				if hasSrcPort {
-					tf.Spec.Packet.TransportHeader.TCP.SrcPort = int32(srcPort)
+		if srcLen > 0 {
+			switch srcType[0] {
+			case "Pod":
+				err := validatePodName(request, src, "source")
+				if err != nil {
+					return nil
 				}
-				if hasDstPort {
-					tf.Spec.Packet.TransportHeader.TCP.DstPort = int32(dstPort)
+				err = validateNamespace(request, srcNamespace, "source")
+				if err != nil {
+					return nil
 				}
-			}
-		case crdv1alpha1.UDPProtocol:
-			{
-				tf.Spec.Packet.TransportHeader.UDP = &crdv1alpha1.UDPHeader{}
-				if hasSrcPort {
-					tf.Spec.Packet.TransportHeader.UDP.SrcPort = int32(srcPort)
+				isSrcPodType = true
+				source = crdv1alpha1.Source{
+					Namespace: srcNamespace,
+					Pod:       src,
 				}
-				if hasDstPort {
-					tf.Spec.Packet.TransportHeader.UDP.DstPort = int32(dstPort)
+			case "IPv4":
+				err := validateIP(request, src, "source")
+				if err != nil {
+					return nil
 				}
-			}
-		case crdv1alpha1.ICMPProtocol:
-			{
-				tf.Spec.Packet.TransportHeader.ICMP = &crdv1alpha1.ICMPEchoRequestHeader{
-					ID:       0,
-					Sequence: 0,
+				source = crdv1alpha1.Source{
+					IP: src,
 				}
 			}
+			sourceName = src
 		}
-		log.Printf("Get user input successfully, traceflow: %+v", tf)
-		tf, err = p.client.CrdV1alpha1().Traceflows().Create(ctx, tf, v1.CreateOptions{})
+		destination := crdv1alpha1.Destination{}
+		if dstLen == 0 && isSrcPodType {
+			goto ContinueWithoutCheckDst
```

Comment:
I do have a question for this part. If we goto ContinueWithoutCheckDst here, do you mean that we will not check duplicate traceflow according to current logics?

Reply:
no, it's not about duplicate traceflow check, for live traffic, empty destination is allowed when source is pod, so it's unnecessary to run destination validation when it's empty.

Reply:
it won't skip destination check when destination is not empty, and it will create a tf as general flow

Reply:
Hi Lan,

Please check my latest question. I may not explain my question well previously. I mean it is not an empty destination or source pod, we will go on check destination. Where do we create tf since it seems that this part of logics is only under ContinueWithoutCheckDst?

Reply:
Hi @mengdie-song ,you can check the logic of GOTO in golang, the code under "ContinueWithoutCheckDst" keyword will be executed as normal code no matter the goto condition is matched or not, the code between "goto ContinueWithoutCheckDst" and "ContinueWithoutCheckDst" will be skipped when the condition matched.

Reply:
I misunderstand the goto usage here and it is only for skipping destination check. Thanks.